drivers: spi: atmel spi_sam0: Reset peripheral on init

pdietl opened the pull request:
I ran into a problem where a vendor-provided bootloader also configured the same SPI peripheral before booting my Zephyr application. The bootloader configured the ENABLE32B option to make writes to the DATA register be 32 bits and not 8 bits -- this broke the Zephyr driver. Since the configuration of the SPI peripheral may be arbitrarily configured before the Zephyr driver attempts to initialize it and because the Zephyr SPI driver assumes that the peripheral is in its default state, it makes sense to use the SWRST bit to ensure that the peripheral is indeed in its default state.

nandojve commented on drivers/spi/spi_sam0.c:
```diff
@@ -657,6 +671,9 @@ static int spi_sam0_init(const struct device *dev)
 			  | GCLK_CLKCTRL_ID(cfg->gclk_id);
 #endif
 
+	regs->CTRLA.bit.SWRST = 1;
+	wait_synchronization(regs);
+
```
I see two things in your PR:
1- Add the software reset at init, which makes sense to me.
2- Handle the 32-bit len extension.

In this case, I think this part should be in their own commit.

pdietl replied:
I will split these changes!

nandojve commented on drivers/spi/spi_sam0.c:
```diff
@@ -77,6 +77,8 @@ static int spi_sam0_configure(const struct device *dev,
 	SercomSpi *regs = cfg->regs;
 	SERCOM_SPI_CTRLA_Type ctrla = {.reg = 0};
 	SERCOM_SPI_CTRLB_Type ctrlb = {.reg = 0};
+	SERCOM_SPI_CTRLC_Type ctrlc = {.reg = 0};
+	SERCOM_SPI_LENGTH_Type length = {.reg = 0};
```
This is a little trick. You need to provide a patch that can be used by the devices that do not have the CTRLC reg.
For instance, this driver should still work with SAMD20.

nandojve reviewed (changes requested):
Overall LGTM, just small things and CI must be green : )

BTW, you need to follow the [commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines).

I'm looking forward to accept you PR.

pdietl commented:
Sweet. Could you let me know what should be improved in the commit message?

pdietl committed:
drivers: spi: sam0: Handle 32-bit length extension

The sam0 SPI driver does not ensure that it clears the 32-bit extension
option during init. The 32-bit extension option, which comprises of a field
in the CTRLC register and the LENGTH register enables better bus
utilization by allowing 32-bit writes to the SPI DATA register
(as opposed to the usual 8-bit writes). The driver breaks down if this
option is enabled by causing each intended byte of output to become
four bytes. We fix this by explicitly disabling the 32-bit extension
option in init.

Signed-off-by: Pete Dietl <petedietl@gmail.com>

pdietl committed:
drivers: spi: sam0: Reset SPI peripheral on init

Reset the SPI peripheral to its default state
and register values on init by setting its SWRST bit.
This is important since the driver assumes that certain
registers are at their default values.

Signed-off-by: Pete Dietl <petedietl@gmail.com>

pdietl commented:
@nandojve I rebased, split the changes into two commits, added conditional compilation for the CTRLC register, and reworded the commit messages. Please let me know if you find this satisfactory or if you desire additional changes :)

pdgendt commented on drivers/spi/spi_sam0.c:
```diff
@@ -129,18 +133,36 @@ static int spi_sam0_configure(const struct device *dev,
 	/* Update the configuration only if it has changed */
-	if (regs->CTRLA.reg != ctrla.reg || regs->CTRLB.reg != ctrlb.reg ||
-	    regs->BAUD.reg != div) {
+	if (regs->CTRLA.reg != ctrla.reg || regs->CTRLB.reg != ctrlb.reg || regs->BAUD.reg != div
+#ifdef SERCOM_SPI_CTRLC_MASK
+		|| regs->LENGTH.reg != length.reg || regs->CTRLC.reg != ctrlc.reg
+#endif
+	) {
```
The first commit has an issue with the opening `(` bracket, which is fixed in the second commit. This results in a non-bisect able history, please fix the first commit.

pdietl replied:
My apologies, very sloppy of me.

pdietl replied:
@pdgendt Done!

pdietl commented:
@pdgendt I think you need to resolve some requested changes.

pdgendt commented:
> @pdgendt I think you need to resolve some requested changes.

How so? All mine are resolved, also note that unresolved comments do not block PRs.

pdietl commented:
@pdgendt ah ok. I just see this 
<img width="674" alt="image" src="https://github.com/user-attachments/assets/1368b591-eceb-41b8-96a8-aa54b2adbccc" />

I guess we need someone with write access to approve?

pdietl commented:
@pdgendt Thanks again for your time with this review!
